lib/rules/jsx-no-literals.js

```diff
@@ -40,116 +167,267 @@ module.exports = {
+      return some(iterFrom([ancestors.parent, ancestors.grandParent]), (parent) => jsxElementTypes.has(parent.type));
+    }
 
-      return parentType === 'JSXFragment' || parentType === 'JSXElement' || grandParentType === 'JSXElement';
+    /**
+     * Determines whether a given node's value and its immediate parent are
+     * viable text nodes that can/should be reported on
+     * @param {ASTNode} node
+     * @param {ResolvedConfig} resolvedConfig
+     * @returns {boolean}
+     */
+    function isViableTextNode(node, resolvedConfig) {
+      const textValues = iterFrom([trimIfString(node.raw), trimIfString(node.value)]);
+      if (some(textValues, (value) => resolvedConfig.allowedStrings.has(value))) {
+        return false;
+      }
+
+      const parent = getParentIgnoringBinaryExpressions(node);
+
+      let isStandardJSXNode = false;
+      if (typeof node.value === 'string' && !/^[\s]+$/.test(node.value) && parent.type.startsWith('JSX')) {
+        if (resolvedConfig.noAttributeStrings) {
+          isStandardJSXNode = parent.type === 'JSXAttribute' || parent.type === 'JSXElement';
+        } else {
+          isStandardJSXNode = parent.type !== 'JSXAttribute';
+        }
+      }
+
+      if (resolvedConfig.noStrings) return isStandardJSXNode;
+      return isStandardJSXNode && parent.type !== 'JSXExpressionContainer';
     }
 
-    function reportLiteralNode(node, messageId) {
-      const ancestorIsJSXElement = hasJSXElementParentOrGrandParent(node);
-      messageId = messageId || defaultMessageId(ancestorIsJSXElement);
+    /**
+     * Gets an override config for a given node. For any given node, we also
+     * need to traverse the ancestor tree to determine if an ancestor's config
+     * will also apply to the current node.
+     * @param {ASTNode} node
+     * @returns {OverrideConfig | undefined}
+     */
+    function getOverrideConfig(node) {
+      if (!hasElementOverrides) return;
+
+      const allAncestorElements = getJSXElementAncestors(node);
+      if (!allAncestorElements.length) return;
+
+      for (const ancestor of allAncestorElements) {
+        const isClosestJSXAncestor = ancestor === allAncestorElements[0];
+
+        const [ancestorName, compoundAncestorName] = getJSXElementName(ancestor);
+        if (ancestorName) {
+          const ancestorConfig = compoundAncestorName
+            ? config.elementOverrides[compoundAncestorName] || config.elementOverrides[ancestorName]
+            : config.elementOverrides[ancestorName];
 
+          if (ancestorConfig) {
+            if (isClosestJSXAncestor || ancestorConfig.applyToNestedElements) {
+              return ancestorConfig;
+            }
+          }
+        }
+      }
+    }
+
+    /**
+     * @param {ResolvedConfig} resolvedConfig
+     * @returns {boolean}
+     */
+    function shouldAllowElement(resolvedConfig) {
+      return resolvedConfig.type === 'override' && 'allowElement' in resolvedConfig && !!resolvedConfig.allowElement;
```

Comment:
`type-check` was being picky here

Reply:
yeah, typescript sadly doesn't infer `x in obj` from a truthy `x.obj`, although it can and should